File: detection/shot_logic.py
"""Trajectory-based make/miss detection.

Ported and cleaned from avishah3/AI-Basketball-Shot-Detection-Tracker
(utils.py + the shot_detector state machine), decoupled from drawing/IO and
reduced to canonical 'ball'/'rim' detections.

Idea: track the ball's recent path. When the ball goes from an 'up' region
(above/around the rim) to a 'down' region (below the rim), count an attempt;
fit a line to the ball's path and see if it crosses the rim plane within the
rim width -> make, else miss.

Each ball/hoop entry is: (center(x, y), frame_idx, w, h, conf)
"""
import math
import numpy as np


def _score(ball_pos, hoop_pos):
    """True if the ball's path crosses the rim plane within the rim width."""
    xs, ys = [], []
    rim_y = hoop_pos[-1][0][1] - 0.5 * hoop_pos[-1][3]
    # first point above the rim + the following point -> a 2-point line
    for i in reversed(range(len(ball_pos))):
        if ball_pos[i][0][1] < rim_y:
            xs.append(ball_pos[i][0][0]); ys.append(ball_pos[i][0][1])
            if i + 1 < len(ball_pos):
                xs.append(ball_pos[i + 1][0][0]); ys.append(ball_pos[i + 1][0][1])
            break
    if len(xs) > 1:
        m, b = np.polyfit(xs, ys, 1)
        if m == 0:
            return False
        pred_x = (rim_y - b) / m
        rim_x1 = hoop_pos[-1][0][0] - 0.4 * hoop_pos[-1][2]
        rim_x2 = hoop_pos[-1][0][0] + 0.4 * hoop_pos[-1][2]
        return rim_x1 < pred_x < rim_x2
    return False


def _detect_down(ball_pos, hoop_pos):
    y = hoop_pos[-1][0][1] + 0.5 * hoop_pos[-1][3]
    return ball_pos[-1][0][1] > y


def _detect_up(ball_pos, hoop_pos):
    h = hoop_pos[-1]
    x1, x2 = h[0][0] - 4 * h[2], h[0][0] + 4 * h[2]
    y1, y2 = h[0][1] - 2 * h[3], h[0][1]
    bx, by = ball_pos[-1][0]
    return x1 < bx < x2 and y1 < by < y2 - 0.5 * h[3]


def _in_hoop_region(center, hoop_pos):
    if not hoop_pos:
        return False
    x, y = center
    h = hoop_pos[-1]
    return (h[0][0] - h[2] < x < h[0][0] + h[2]) and (h[0][1] - h[3] < y < h[0][1] + 0.5 * h[3])


def _clean_ball(ball_pos, frame_idx, max_frames):
    """Drop points that jump too far (wrong ball) or aren't roughly square."""
    if len(ball_pos) > 1:
        w1, h1 = ball_pos[-2][2], ball_pos[-2][3]
        (x1, y1), (x2, y2) = ball_pos[-2][0], ball_pos[-1][0]
        f_dif = ball_pos[-1][1] - ball_pos[-2][1]
        if math.hypot(x2 - x1, y2 - y1) > 4 * math.hypot(w1, h1) and f_dif < 5:
            ball_pos.pop()
        else:
            w2, h2 = ball_pos[-1][2], ball_pos[-1][3]
            if w2 * 1.4 < h2 or h2 * 1.4 < w2:
                ball_pos.pop()
    if ball_pos and frame_idx - ball_pos[0][1] > max_frames:
        ball_pos.pop(0)
    return ball_pos


def _clean_hoop(hoop_pos):
    if len(hoop_pos) > 1:
        (x1, y1), (x2, y2) = hoop_pos[-2][0], hoop_pos[-1][0]
        w1, h1 = hoop_pos[-2][2], hoop_pos[-2][3]
        f_dif = hoop_pos[-1][1] - hoop_pos[-2][1]
        if math.hypot(x2 - x1, y2 - y1) > 0.5 * math.hypot(w1, h1) and f_dif < 5:
            hoop_pos.pop()
        else:
            w2, h2 = hoop_pos[-1][2], hoop_pos[-1][3]
            if w2 * 1.3 < h2 or h2 * 1.3 < w2:
                hoop_pos.pop()
    if len(hoop_pos) > 25:
        hoop_pos.pop(0)
    return hoop_pos


class ShotTracker:
    """Feed canonical detections per frame; get a make/miss event on a completed shot."""

    def __init__(self, max_frames=30, ball_conf=0.35, rim_conf=0.45):
        self.ball_pos, self.hoop_pos = [], []
        self.max_frames = max_frames
        self.ball_conf, self.rim_conf = ball_conf, rim_conf
        self.up = self.down = False
        self.up_frame = self.down_frame = 0
        self.makes = self.attempts = 0
        self.frame = 0

    def update(self, dets, frame_idx=None):
        """dets: list of (kind, (cx, cy), w, h, conf) with kind in {'ball', 'rim'}.
        Returns an event dict on a completed attempt, else None."""
        frame_idx = self.frame if frame_idx is None else frame_idx
        self.frame = frame_idx
        for kind, center, w, h, conf in dets:
            if kind == "ball" and (conf > self.ball_conf or
                                   (_in_hoop_region(center, self.hoop_pos) and conf > 0.15)):
                self.ball_pos.append((center, frame_idx, w, h, conf))
            elif kind == "rim" and conf > self.rim_conf:
                self.hoop_pos.append((center, frame_idx, w, h, conf))

        self.ball_pos = _clean_ball(self.ball_pos, frame_idx, self.max_frames)
        if len(self.hoop_pos) > 1:
            self.hoop_pos = _clean_hoop(self.hoop_pos)
        return self._detect(frame_idx)

    def _detect(self, frame_idx):
        if not self.hoop_pos or not self.ball_pos:
            return None
        if not self.up:
            self.up = _detect_up(self.ball_pos, self.hoop_pos)
            if self.up:
                self.up_frame = self.ball_pos[-1][1]
        if self.up and not self.down:
            self.down = _detect_down(self.ball_pos, self.hoop_pos)
            if self.down:
                self.down_frame = self.ball_pos[-1][1]

        if self.up and self.down and self.up_frame < self.down_frame:
            self.attempts += 1
            made = _score(self.ball_pos, self.hoop_pos)
            if made:
                self.makes += 1
            self.up = self.down = False
            return {
                "result": "make" if made else "miss",
                "attempt": self.attempts,
                "makes": self.makes,
                "frame": frame_idx,
                "up_frame": self.up_frame,
                "ball_path": [tuple(map(int, p[0])) for p in self.ball_pos],
            }
        return None

    @property
    def fg_pct(self):
        return 100.0 * self.makes / self.attempts if self.attempts else 0.0

    @property
    def rim_center(self):
        return self.hoop_pos[-1][0] if self.hoop_pos else None

File: mobile/www/shot_logic.js
/* HoopTracker mobile - trajectory make/miss.
   Faithful port of detection/shot_logic.py. Loaded as a plain <script>;
   exposes window.HoopShot.ShotTracker.
   Detection objects passed to update(): { kind:'ball'|'rim', c:[x,y], w, h, conf } */
(function (global) {
  function hypot(a, b) { return Math.sqrt(a * a + b * b); }

  function polyfit1(xs, ys) {
    const n = xs.length; let sx = 0, sy = 0, sxx = 0, sxy = 0;
    for (let i = 0; i < n; i++) { sx += xs[i]; sy += ys[i]; sxx += xs[i] * xs[i]; sxy += xs[i] * ys[i]; }
    const d = n * sxx - sx * sx; if (Math.abs(d) < 1e-9) return [0, 0];
    const m = (n * sxy - sx * sy) / d; return [m, (sy - m * sx) / n];
  }

  function score(ball, hoop) {
    const h = hoop[hoop.length - 1];
    const rimY = h.c[1] - 0.5 * h.h;
    const xs = [], ys = [];
    for (let i = ball.length - 1; i >= 0; i--) {
      if (ball[i].c[1] < rimY) {
        xs.push(ball[i].c[0]); ys.push(ball[i].c[1]);
        if (i + 1 < ball.length) { xs.push(ball[i + 1].c[0]); ys.push(ball[i + 1].c[1]); }
        break;
      }
    }
    if (xs.length > 1) {
      const fit = polyfit1(xs, ys); const m = fit[0], b = fit[1];
      if (m === 0) return false;
      const px = (rimY - b) / m;
      return (h.c[0] - 0.4 * h.w) < px && px < (h.c[0] + 0.4 * h.w);
    }
    return false;
  }

  function detectDown(ball, hoop) {
    const h = hoop[hoop.length - 1];
    return ball[ball.length - 1].c[1] > h.c[1] + 0.5 * h.h;
  }

  function detectUp(ball, hoop) {
    const h = hoop[hoop.length - 1];
    const x1 = h.c[0] - 4 * h.w, x2 = h.c[0] + 4 * h.w;
    const y1 = h.c[1] - 2 * h.h, y2 = h.c[1];
    const b = ball[ball.length - 1].c;
    return x1 < b[0] && b[0] < x2 && y1 < b[1] && b[1] < (y2 - 0.5 * h.h);
  }

  function inHoopRegion(c, hoop) {
    if (!hoop.length) return false;
    const h = hoop[hoop.length - 1];
    return (h.c[0] - h.w) < c[0] && c[0] < (h.c[0] + h.w) &&
           (h.c[1] - h.h) < c[1] && c[1] < (h.c[1] + 0.5 * h.h);
  }

  function cleanBall(ball, frameIdx, maxFrames) {
    if (ball.length > 1) {
      const a = ball[ball.length - 2], b = ball[ball.length - 1];
      const dist = hypot(b.c[0] - a.c[0], b.c[1] - a.c[1]);
      if (dist > 4 * hypot(a.w, a.h) && (b.f - a.f) < 5) ball.pop();
      else if (b.w * 1.4 < b.h || b.h * 1.4 < b.w) ball.pop();
    }
    if (ball.length && (frameIdx - ball[0].f) > maxFrames) ball.shift();
    return ball;
  }

  function cleanHoop(hoop) {
    if (hoop.length > 1) {
      const a = hoop[hoop.length - 2], b = hoop[hoop.length - 1];
      const dist = hypot(b.c[0] - a.c[0], b.c[1] - a.c[1]);
      if (dist > 0.5 * hypot(a.w, a.h) && (b.f - a.f) < 5) hoop.pop();
      else if (b.w * 1.3 < b.h || b.h * 1.3 < b.w) hoop.pop();
    }
    if (hoop.length > 25) hoop.shift();
    return hoop;
  }

  class ShotTracker {
    constructor(opts) {
      opts = opts || {};
      this.maxFrames = opts.maxFrames || 30;
      this.ballConf = opts.ballConf || 0.35;
      this.rimConf = opts.rimConf || 0.45;
      this.ball = []; this.hoop = [];
      this.up = false; this.down = false; this.upFrame = 0; this.downFrame = 0;
      this.makes = 0; this.attempts = 0; this.frame = 0;
    }
    get fgPct() { return this.attempts ? 100 * this.makes / this.attempts : 0; }
    get rimCenter() { return this.hoop.length ? this.hoop[this.hoop.length - 1].c : null; }

    update(dets, frameIdx) {
      this.frame = frameIdx;
      for (const d of dets) {
        if (d.kind === 'ball' && (d.conf > this.ballConf || (inHoopRegion(d.c, this.hoop) && d.conf > 0.15)))
          this.ball.push({ c: d.c, f: frameIdx, w: d.w, h: d.h, conf: d.conf });
        else if (d.kind === 'rim' && d.conf > this.rimConf)
          this.hoop.push({ c: d.c, f: frameIdx, w: d.w, h: d.h, conf: d.conf });
      }
      cleanBall(this.ball, frameIdx, this.maxFrames);
      if (this.hoop.length > 1) cleanHoop(this.hoop);
      return this._detect(frameIdx);
    }

    _detect(frameIdx) {
      if (!this.hoop.length || !this.ball.length) return null;
      if (!this.up) { this.up = detectUp(this.ball, this.hoop); if (this.up) this.upFrame = this.ball[this.ball.length - 1].f; }
      if (this.up && !this.down) { this.down = detectDown(this.ball, this.hoop); if (this.down) this.downFrame = this.ball[this.ball.length - 1].f; }
      if (this.up && this.down && this.upFrame < this.downFrame) {
        this.attempts++;
        const made = score(this.ball, this.hoop);
        if (made) this.makes++;
        this.up = false; this.down = false;
        return { result: made ? 'make' : 'miss', attempt: this.attempts, makes: this.makes, frame: frameIdx };
      }
      return null;
    }
  }

  global.HoopShot = { ShotTracker };
})(window);
